src/recon.rs

```diff
@@ -535,30 +542,81 @@ fn decode_coefs<BD: BitDepth>(
+        S32x32 => decode_coefs_inner::<BD, { S32x32 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        S64x64 => decode_coefs_inner::<BD, { S64x64 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R4x8 => decode_coefs_inner::<BD, { R4x8 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R8x4 => decode_coefs_inner::<BD, { R8x4 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R8x16 => decode_coefs_inner::<BD, { R8x16 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R16x8 => decode_coefs_inner::<BD, { R16x8 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R16x32 => decode_coefs_inner::<BD, { R16x32 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R32x16 => decode_coefs_inner::<BD, { R32x16 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R32x64 => decode_coefs_inner::<BD, { R32x64 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R64x32 => decode_coefs_inner::<BD, { R64x32 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R4x16 => decode_coefs_inner::<BD, { R4x16 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R16x4 => decode_coefs_inner::<BD, { R16x4 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R8x32 => decode_coefs_inner::<BD, { R8x32 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R32x8 => decode_coefs_inner::<BD, { R32x8 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R16x64 => decode_coefs_inner::<BD, { R16x64 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+        R64x16 => decode_coefs_inner::<BD, { R64x16 as _ }>(f, ts, ts_c, dbg_block_info, scratch, t_cf, a, l, tx, bs, b, plane, cf, txtp, res_ctx),
+    }
+}
+
+#[inline(never)]
+fn decode_coefs_inner<BD: BitDepth, const txsize_discr: usize>(
+    f: &Rav1dFrameData,
+    ts: usize,
+    ts_c: &mut Rav1dTileStateContext,
+    dbg_block_info: bool,
+    scratch: &mut TaskContextScratch,
+    t_cf: &mut Cf,
+    a: &mut [u8],
+    l: &mut [u8],
+    tx: TxfmSize,
+    bs: BlockSize,
+    b: &Av1Block,
+    plane: usize,
+    cf: CfSelect,
+    txtp: &mut TxfmType,
+    res_ctx: &mut u8,
+) -> c_int {
+    macro_rules! tdim {
+        () => {
+            dav1d_txfm_dimension::<txsize_discr>()
```

Comment:
Does `const TDIM: TxfmInfo = dav1d_txfm_dimension::<txsize_discr>();` or `let tdim = const { dav1d_txfm_dimension::<txsize_discr>() };` work, too?

Reply:
The first won't but the second should yeah. As I said above I haven't tried to see how it'd codegen just yet. I'll do so soon. 

But there are many knobs we can tweak in how we accomplish the goal of pushing down constant values. I believe we can indeed clean this up, and also wished to compare to how Nick did it with the 8/9 const generics.

Reply:
I've put my rough code at #1326.

Reply:
> Does `const TDIM: TxfmInfo = dav1d_txfm_dimension::<txsize_discr>();` or `let tdim = const { dav1d_txfm_dimension::<txsize_discr>() };` work, too?

so I did try the latter as well, and tried to check the many monomorphizations' codegen:
- codegen didn't meaningfully change for `get_skip_ctx` and `get_dc_sign_ctx` (which has a similar var)
- most of the `decode_coefs` monomorphizations get completely shuffled around by the change, making comparisons much harder on this big function. Most of what I looked at seemed correctly optimized out but it's hard to tell for sure it was better/not worse. Additionally, measurements seemed slightly worse again, but so close to noise on the 2 CPUs I tried that I am not 100% sure. So I didn't add this change to the PR, but can push to a different branch if you'd also like to test that.

Reply:
I'd like to benchmark the const assignment, if you have that change handy.

Reply:
Absolutely, [here's](https://github.com/lqd/rav1d/commit/09e1e2590f669497fc4b6bb9f16ae540b2a418ee) the commit.

Rerunning a quick benchmark now, it doesn't look as different as I saw earlier, but it's still a bit noisy. (The difference may have been when I looked at a non-const generic const fn, but the optimization pipeline should see the same MIR :/). It looked slightly negative on 7502P and slightly positive on the 13900KF.

Reply:
They are equivalent in my testing, variance is within noise.

Reply:
Good, thank you for checking. I'll push that commit here then.